File: ModuleDaily/src/main/java/com/qihuan/dailymodule/presenter/DailyDetPresenter.java
package com.qihuan.dailymodule.presenter;

import com.qihuan.commonmodule.base.BasePresenterImpl;
import com.qihuan.commonmodule.collection.bean.CollectionBean;
import com.qihuan.commonmodule.collection.model.CollectionModel;
import com.qihuan.commonmodule.net.ApiManager;
import com.qihuan.dailymodule.contract.DailyDetContract;
import com.qihuan.dailymodule.model.ZhihuApi;
import com.qihuan.dailymodule.model.bean.StoryContentBean;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class DailyDetPresenter extends BasePresenterImpl<DailyDetContract.View> implements DailyDetContract.Presenter {

    private final CollectionModel collectionModel;
    private StoryContentBean storyContentBean;

    public DailyDetPresenter() {
        collectionModel = new CollectionModel();
    }

    @Override
    public void detachView() {
        super.detachView();
        collectionModel.onDestroy();
    }

    @Override
    public void getStoryContent(int id) {
        addDisposable(
                ApiManager.getInstance()
                        .getApi(ZhihuApi.class)
                        .getStoryContent(id)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(storyContentBean -> {
                            if (storyContentBean == null) {
                                getView().showError("该日报已不存在~");
                                return;
                            }
                            this.storyContentBean = storyContentBean;
                            getView().storyContent(storyContentBean);
                        }, throwable -> getView().showError(throwable.getMessage()))
        );
    }

    @Override
    public void getStoryExtra(int id) {
        addDisposable(
                ApiManager.getInstance()
                        .getApi(ZhihuApi.class)
                        .getStoryExtra(id)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(storyExtraBean -> getView().storyExtra(storyExtraBean),
                                throwable -> getView().showError(throwable.getMessage()))
        );
    }

    @Override
    public void getFavoriteStory(int id) {
        getView().showLoading();
        collectionModel.getFavoriteList(String.valueOf(id), collectionList -> {
            getView().onFavoriteChange(collectionList.size() != 0);
            getView().hideLoading();
        });
    }

    @Override
    public void updateFavoriteStory(int id) {
        if (storyContentBean == null) {
            return;
        }
        CollectionBean collectionBean = new CollectionBean()
                .setCollectionId(String.valueOf(id))
                .setType(0)
                .setTitle(storyContentBean.getTitle())
                .setImg(storyContentBean.getImage());
        collectionModel.updateFavorite(collectionBean, isFavorite -> {
            getView().hideLoading();
            getView().onFavoriteChange(isFavorite);
            getView().showUpdateFavoriteInfo(isFavorite);
        });
    }

}

File: ModuleDaily/src/main/java/com/qihuan/dailymodule/presenter/DailyDetPresenter.kt
package com.qihuan.dailymodule.presenter

import android.util.Log
import com.qihuan.commonmodule.base.AbsRxPresenter
import com.qihuan.commonmodule.collection.CollectionBean
import com.qihuan.commonmodule.db.AppDatabase
import com.qihuan.dailymodule.contract.DailyDetContract
import com.qihuan.dailymodule.model.ZhihuApi
import com.qihuan.dailymodule.model.bean.StoryContentBean
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers

class DailyDetPresenter : AbsRxPresenter<DailyDetContract.View>(), DailyDetContract.Presenter {

    private var storyContentBean: StoryContentBean? = null

    override fun getStoryContent(id: Int) {
        ZhihuApi.get()
                .getStoryContent(id)
                .doOnSubscribe { addDisposable(it) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                        onNext = {
                            this.storyContentBean = it
                            view?.storyContent(it)
                        },
                        onError = {
                            view?.showError(it.message ?: "")
                        }
                )
    }

    override fun getStoryExtra(id: Int) {
        ZhihuApi.get()
                .getStoryExtra(id)
                .doOnSubscribe { addDisposable(it) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                        onNext = {
                            view?.storyExtra(it)
                        },
                        onError = {
                            view?.showError(it.message ?: "")
                        }
                )
    }

    override fun getFavoriteStory(id: Int) {
        AppDatabase.instance
                .collectionDao()
                .queryOne(id.toString())
                .doOnSubscribe { addDisposable(it) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                        onSuccess = {
                            view?.onFavoriteChange(true)
                        },
                        onError = {
                            view?.onFavoriteChange(false)
                        }
                )
    }

    override fun updateFavoriteStory(id: Int) {
        AppDatabase.instance
                .collectionDao()
                .queryOne(id.toString())
                .doOnSubscribe { addDisposable(it) }
                .flatMapCompletable {
                    return@flatMapCompletable Completable.fromAction { AppDatabase.instance.collectionDao().delete(it) }
                }
                .onErrorResumeNext {
                    return@onErrorResumeNext Completable.fromAction {
                        AppDatabase.instance.collectionDao().save(CollectionBean(id.toString(), title = storyContentBean?.title, img = storyContentBean?.image))
                    }
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                        onComplete = {
                            Log.e("updateFavoriteStory", "success")
                        },
                        onError = {
                            Log.e("updateFavoriteStory", "onError", it)
                        }
                )
    }

}
